shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm

```diff
@@ -75,24 +48,64 @@ - (instancetype)initWithTaskRunner:(fml::RefPtr<fml::TaskRunner>)task_runner
 
   if (self) {
     callback_ = std::move(callback);
-    display_link_ = fml::scoped_nsobject<CADisplayLink> {
-      [[CADisplayLink displayLinkWithTarget:self selector:@selector(onDisplayLink:)] retain]
-    };
-    display_link_.get().paused = YES;
+    display_link_ = [CADisplayLink displayLinkWithTarget:self selector:@selector(onDisplayLink:)];
+    display_link_.paused = YES;
 
     task_runner->PostTask([client = [self retain]]() {
-      [client->display_link_.get() addToRunLoop:[NSRunLoop currentRunLoop]
-                                        forMode:NSRunLoopCommonModes];
+      [client->display_link_ addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
       [client release];
     });
   }
 
   return self;
 }
 
-- (float)displayRefreshRate {
+- (void)await {
+  display_link_.paused = NO;
+}
+
+- (void)onDisplayLink:(CADisplayLink*)link {
+  TRACE_EVENT0("flutter", "VSYNC");
+
+  CFTimeInterval delay = CACurrentMediaTime() - link.timestamp;
+  fml::TimePoint frame_start_time = fml::TimePoint::Now() - fml::TimeDelta::FromSecondsF(delay);
+  fml::TimePoint frame_target_time = frame_start_time + fml::TimeDelta::FromSecondsF(link.duration);
+
+  display_link_.paused = YES;
+
+  callback_(frame_start_time, frame_target_time);
+}
+
+- (void)invalidate {
+  [display_link_ invalidate];
+}
+
+- (void)dealloc {
+  [self invalidate];
+
```

Comment:
Shouldn't we dealloc `display_link_` here now?

Reply:
Yup and done!

Reply:
Seems like calling [invalidate](https://developer.apple.com/documentation/quartzcore/cadisplaylink/1621293-invalidate?language=objc) does release all strongly held references of the display_link object (from all the run loops). I've also seen this in local testing.

[This SO post](https://stackoverflow.com/questions/47368609/definitively-do-you-have-to-invalidate-a-cadisplaylink-when-the-controller-di#:~:text=Method%20definition%20of%20invalidate(),and%20run%20loop%20holds%20DispayLink.) also indicates the same, I don't think we need to `dealloc` it.